# Optimization Problems

In [4]:
import numpy as np 
import sympy as sym
from scipy.optimize import linprog
from scipy.optimize import minimize

# Problem 1

### Problem statement

A certain corporation has three branch plants with excess production capacity. The three plants have the necessary elements to produce a certain product, and the manager has decided to use part of the excess production capacity for this purpose. This product can be made in three sizes: large, medium, and small, which result in a net profit of $140, $120, and $100, respectively.

Plants 1, 2, and 3 have excess labor and equipment capacity to produce 750, 800, and 450 units per day of this product, respectively, regardless of size or combination of sizes applied. However, the available storage space for these products also imposes a limitation on production rates. Plants 1, 2, and 3 have 13,000, 12,000, and 5,000 square feet of possible storage space for in-process products, for one day of production of this item. Each unit of the large, medium, and small sizes produced per day requires 20, 15, and 12 square feet, respectively.

Sales forecasts indicate that at most 900, 1,200, and 750 units of the large, medium, and small sizes, respectively, can be sold. The manager hires you to determine how much should be produced of each size at each plant to maximize profit.


## a) 
The decision variables are the size of the products: large, medium, and small. These are represented through the variables $x_1$ (large), $x_2$ (medium), and $x_3$ (small).

The notation $x_{ij}$ is defined to express the products belonging to plant $i$, of size $j$.

Additionally, the quantity of products ($x_j$) is defined by the following equations:

1) $x_1 = x_{11} + x_{21} + x_{31}$ (Large)

2) $x_2 = x_{12} + x_{22} + x_{32}$ (Medium)

3) $x_3 = x_{13} + x_{23} + x_{33}$ (Small)


## b) 
The objective function is given by $f(x)$.

$f(x) = 140x_1 + 120x_2 + 100x_3$ \
$f(x) = 140(x_{11} + x_{21} + x_{31}) + 120(x_{12} + x_{22} + x_{32}) + 100(x_{13} + x_{23} + x_{33})$ \
$f(x) = 140x_{11} + 140x_{21} + 140x_{31} +  120x_{12} + 120x_{22} + 120x_{32} + 100x_{13} + 100x_{23} + 100x_{33}$


### Here are the space constraints in the plants:

Plant 1:

$20x_{11} + 15x_{12} + 12x_{13} \leq 13000$

Plant 2:

$20x_{21} + 15x_{22} + 12x_{23} \leq 12000$

Plant 3:

$20x_{31} + 15x_{32} + 12x_{33} \leq 5000$

### Here are the production capacity constraints of the plants:

Plant 1:

$x_{11} + x_{12} + x_{13} \leq 750$

Plant 2:

$x_{21} + x_{22} + x_{23} \leq 900$

Plant 3:

$x_{31} + x_{32} + x_{33} \leq 450$

### Here are the sales forecast constraints:

Large:

$x_1 = x_{11} + x_{21} + x_{31} \leq 900$

Medium:

$x_2 = x_{12} + x_{22} + x_{32} \leq 1200$

Small:

$x_3 = x_{13} + x_{23} + x_{33} \leq 750$


## c) Optimization problem solution


Negative values ​​of the vector are placed because the negative of the objective function is being minimized.

In [5]:
c = np.array([-140, -120, -100, -140, -120, -100, -140, -120, -100])
A = np.array([[20,15,12,0,0,0,0,0,0], [0,0,0,20,15,12,0,0,0], [0,0,0,0,0,0,20,15,12], [1,1,1,0,0,0,0,0,0], [0,0,0,1,1,1,0,0,0], [0,0,0,0,0,0,1,1,1], [1,0,0,1,0,0,1,0,0], [0,1,0,0,1,0,0,1,0], [0,0,1,0,0,1,0,0,1]])
b = np.array([13000, 12000, 5000, 750, 900, 450, 900, 1200, 750])
x11_bounds = (0, None)
x12_bounds = (0, None)
x13_bounds = (0, None)
x21_bounds = (0, None)
x22_bounds = (0, None)
x23_bounds = (0, None)
x31_bounds = (0, None)
x32_bounds = (0, None)
x33_bounds = (0, None)

res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x11_bounds, x12_bounds, x13_bounds, x21_bounds, x22_bounds, x23_bounds, x31_bounds, x32_bounds, x33_bounds]), method='revised simplex', options={"disp": True})
print(res)

Phase Iteration Minimum Slack       Constraint Residual Objective          
1     0         450.0               0.0                 0.0                 
Phase Iteration Minimum Slack       Constraint Residual Objective          
2     0         450.0               0.0                 0.0                 
2     1         0.0                 0.0                 -91000.0            
2     2         0.0                 0.0                 -126000.0           
2     3         0.0                 0.0                 -174000.0           
2     4         0.0                 0.0                 -182000.0           
2     5         0.0                 0.0                 -222000.0           
2     6         0.0                 0.0                 -228000.0           
2     7         0.0                 0.0                 -234666.6666667     
2     8         0.0                 0.0                 -236000.0           
Optimization terminated successfully.
         Current function value: -236000

/var/folders/m8/d4hdb3fs173cl7w2mxqmt8lc0000gn/T/ipykernel_50808/2821817496.py:14: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x11_bounds, x12_bounds, x13_bounds, x21_bounds, x22_bounds, x23_bounds, x31_bounds, x32_bounds, x33_bounds]), method='revised simplex', options={"disp": True})


### Objective function value

In [6]:
print(-1*res.fun)

235999.99999999997


### Value of each variable

In [7]:
print(res.x)

[350.         400.           0.           0.         533.33333333
 333.33333333   0.           0.         416.66666667]


### The production should be:

**350** products of **large** size in plant 1 ($x_{11} = 350$)

**400** products of **medium** size in plant 1 ($x_{12} = 400$)

**0** products of **small** size in plant 1 ($x_{13} = 0$)

### Giving a total production in plant 1 of

$x_1 = x_{11} + x_{12} + x_{13} = 750$ products

---

### The production should be:

**0** products of **large** size in plant 2 ($x_{21} = 0$)

**533.33** products of **medium** size in plant 2 ($x_{22} = 533.33$)

**333.33** products of **small** size in plant 2 ($x_{23} = 333.33$)

### Giving a total production in plant 2 of

$x_2 = x_{21} + x_{22} + x_{23} = 866.66$ products

*Note: Since in reality an irrational number of products cannot be produced, this value would be rounded to the nearest integer, which in this case would be 867.*

---

### The production should be:

**0** products of **large** size in plant 3 ($x_{31} = 0$)

**0** products of **medium** size in plant 3 ($x_{32} = 0$)

**416.66** products of **small** size in plant 3 ($x_{33} = 416.66$)

### Giving a total production in plant 3 of

$x_3 = x_{31} + x_{32} + x_{33} = 416.66$ products

*Note: Since in reality an irrational number of products cannot be produced, this value would be rounded to the nearest integer, which in this case would be 417.*

---


**These results guarantee that the profit is maximized at a value of $f(x) = \$236,000$**

# Problem 2

### Problem Statement

A store manager wishes to minimize the distance traveled by his transport trucks from the three stores he has across town, to the warehouse where he keeps his inventory. It is known that the warehouse and store **A** lie at the origin of the cartesian plane $(0,0)$, that store **B** lies in the point $(40, 50)$ of the plane, and that store **C** lies in the point $(90, 30)$ of the plane. 

He wants to know what the optimal location for the warehouse is, which minimizes the mean distance from the warehouse, to each store, and hires you to solve this problem for him.

### a) Decision variables

$x$ position (horizontal) \
$y$ position (vertical) \
of the warehouse on a 2D plane

### b) Objective function
Distance between store A and the warehouse: $d_{a} = \sqrt{x^{2}+y^{2}}$

Distance between store B and the warehouse: $d_{b} = \sqrt{(x-40)^{2}+(y-50)^{2}}$

Distance between store C and the warehouse: $d_{c} = \sqrt{(x-90)^{2}+(y-30)^{2}}$

Function to optimize: $f(x,y) = d_a + d_b + d_c$


In [8]:
def f(x):
    return sym.sqrt(x[0]**2+x[1]**2) + sym.sqrt((x[0]-40)**2 + (x[1]-50)**2) + sym.sqrt((x[0]-90)**2+(x[1]-30)**2)

### c) Can this problem be reformulated linearly without changing the decision variables?

This problem cannot be reformulated linearly without changing the decision variables because achieving this would require a variable change involving a complex mathematical transformation to eliminate the square roots, which still does not guarantee the linearity of the problem. Additionally, doing so implies a change in the decision variables.

### d) Optimization using the Broyden-Fletcher-Goldfarb-Shanno method

In [9]:
# We define a vector of initial values for the decision variables.
x0 = [0,0]

res = minimize(f, x0, method='bfgs', options={'disp': True})
print(res)
print(np.round(res.x,3))

Optimization terminated successfully.
         Current function value: 117.114336
         Iterations: 11
         Function evaluations: 48
         Gradient evaluations: 16
  message: Optimization terminated successfully.
  success: True
   status: 0
      fun: 117.11433586454925
        x: [ 4.204e+01  4.255e+01]
      nit: 11
      jac: [ 3.815e-06  0.000e+00]
 hess_inv: [[ 9.291e+00 -7.635e+00]
            [-7.635e+00  3.389e+01]]
     nfev: 48
     njev: 16
[42.045 42.546]


### e)

According to the result obtained in section $d)$, it can be stated that the optimal location for the warehouse (closest possible to stores A, B, and C) corresponds to the coordinates ($x$, $y$) = (42.045, 42.546).

# Problem 3

### Problem Statement

The store manager now asks you what the optimal location of the warehouses should be if he now introduces a second warehouse. That is, there would now be one warehouse to supply store **A** and **B**, and a second warehouse to supply stores **B** and **C**.

### a) Expresión matemática

Distance between store A and its respective warehouse: $d_{a} = \sqrt{x^{2}+y^{2}}$ 

Distance between store B and its respective warehouse: $d_{b} = \sqrt{(x-40)^{2}+(y-50)^{2}}$ 

Distance between store C and its respective warehouse: $d_{c} = \sqrt{(x-90)^{2}+(y-30)^{2}}$

Variables $x_1$ and $y_1$ are defined to represent the coordinates of the warehouse corresponding to store A.

Variables $x_2$ and $y_2$ are defined to represent the coordinates of the warehouse corresponding to store C.

Objective function:

$f(x_1,x_2,y_1,y_2) = \sqrt{x_{1}^{2}+y_{1}^{2}} + \frac{1}{2} \left(\sqrt{(x_{1}-40)^{2}+(y_{1}-50)^{2}} + \sqrt{(x_{2}-40)^{2}+(y_{2}-50)^{2}}\right) + \sqrt{(x_{2}-90)^{2} + (y_{2}-30)^{2}} $

In [10]:
# x is a vector of 4 components. Each component represents one of the variables of the objective function.
def f1(x):
    return np.sqrt(x[0]**2+x[1]**2) + (1/2)*(np.sqrt((x[0]-40)**2 + (x[1]-50)**2) + np.sqrt((x[2]-40)**2 + (x[3]-50)**2)) + np.sqrt((x[2]-90)**2+(x[3]-30)**2)
    

### b) Minimization of the Objective Function

Information on how to use the *minimize* function was found [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html).

In [11]:
x1 = [0,0,0,0]

res2 = minimize(f1, x1, method='BFGS', options={'disp': True})
print(res2)

         Current function value: 58.941445
         Iterations: 66
         Function evaluations: 912
         Gradient evaluations: 180
  message: Desired error not necessarily achieved due to precision loss.
  success: False
   status: 2
      fun: 58.94144524135208
        x: [-1.246e-09 -4.023e-09  9.000e+01  3.000e+01]
      nit: 66
      jac: [ 3.603e-01  6.173e-02  3.490e-01 -5.023e-01]
 hess_inv: [[ 6.262e-08  4.248e-09  8.255e-09 -1.307e-08]
            [ 4.248e-09  1.490e-08  2.939e-09 -9.667e-09]
            [ 8.255e-09  2.939e-09  1.153e-08 -6.322e-09]
            [-1.307e-08 -9.667e-09 -6.322e-09  2.366e-08]]
     nfev: 912
     njev: 180


/Users/juanpablonaranjo/anaconda3/lib/python3.11/site-packages/scipy/optimize/_minimize.py:705: OptimizeWarning: Desired error not necessarily achieved due to precision loss.
  res = _minimize_bfgs(fun, x0, args, jac, callback, **options)


The result obtained for the first two coordinates (x1, y1) are numbers in the order of $10^{-9}$ and $10^{-10}$, which means that those numbers are basically 0.

As can be noted, the obtained response for the location of the warehouses are respectively in the locations of the stores A and C, that is, at coordinates (0, 0) for store A, and at coordinates (90, 30) for store C.

In [12]:
res21 = minimize(f1, x1, method='Nelder-Mead', options={'disp': True})
print(res21)

Optimization terminated successfully.
         Current function value: 82.696593
         Iterations: 216
         Function evaluations: 360
       message: Optimization terminated successfully.
       success: True
        status: 0
           fun: 82.6965929183624
             x: [-1.423e+01 -7.694e+00  9.000e+01  3.000e+01]
           nit: 216
          nfev: 360
 final_simplex: (array([[-1.423e+01, -7.694e+00,  9.000e+01,  3.000e+01],
                       [-1.423e+01, -7.694e+00,  9.000e+01,  3.000e+01],
                       ...,
                       [-1.423e+01, -7.694e+00,  9.000e+01,  3.000e+01],
                       [-1.423e+01, -7.694e+00,  9.000e+01,  3.000e+01]]), array([ 8.270e+01,  8.270e+01,  8.270e+01,  8.270e+01,
                        8.270e+01]))


In this case, the Nelder-Mead method was used instead of the BFGS method to carry out the minimization. As can be noted, the result obtained for the coordinates of the first warehouse (x1, y1) are very different from those obtained by the BFGS method. Meanwhile, the coordinates of the second warehouse (x2, y2) are similar to those obtained by the BFGS method, being around the point (90, 30) again.

This may be due to the fact that the [Nelder-Mead algorithm](http://www.scholarpedia.org/article/Nelder-Mead_algorithm) only uses values of the objective function at some points in $\mathbb{R}^n$ and does not calculate the gradient of the objective function at these points to move towards the optimal solution, as the [BFGS method](https://en.wikipedia.org/wiki/Broyden%E2%80%93Fletcher%E2%80%93Goldfarb%E2%80%93Shanno_algorithm) does. This could be causing the discrepancy between the values obtained for the coordinates (x1, y1) between the two analyzed methods.

### c) According to the previous results, which of the two options (between 1 and 2 warehouses) would you recommend to the store manager?

Based on the results obtained in problem 2 and in section *b)* of this problem, it can be stated that it is more advisable to have two separate warehouses (one for location A and one for C, distributing the production of location B) because in this way, the optimal location for both warehouses is right next to their respective location, as it is the shortest distance possible between each one and its respective location.

With a single warehouse for all three locations, an optimal warehouse location is obtained just at a relatively central point for the three locations, although it is clear that the distance to location A is farther than the distance to the other two locations respectively.